cloudbot/util/decorators.py

```diff
@@ -0,0 +1,25 @@
+from time import time
+
+SPAM_LIST = {}
+
+
+def spam_guard(func):
+    def wrapper(*args, **kwargs):
+        nick = kwargs.get('nick', None)  # pull the nick from the args
+        if nick:  # if there is no nick specified for whatever reason, carry on
+            global SPAM_LIST
+            if SPAM_LIST.get(nick, None):
+                SPAM_LIST[nick][0] = time()
+                if time() - SPAM_LIST[nick][0] < SPAM_LIST[nick][1]:
+                    # notice that they are spamming
+                    SPAM_LIST[nick][1] += 10
+                    return (
+                        "{} attempted to execute a command within their previous" \
+                        " limit. An additional 10 seconds has been added. Current" \
+                        " limit is {} seconds.".format(nick, SPAM_LIST[nick][1])
```

Comment:
I see a few problems with this: 

1. The user doesn't necessarily know they've fallen afoul of the spam guard (unless the return causes the bot to inform the whole channel/medium by which the command was attempted?)
2. The only remediation is reloading the whole bot, which has problems of its own:  
  a. Other users that have fallen afoul of the spam guard will suddenly be "redeemed"  
  b. Full saves will be required for all other plugins before reload, otherwise all changes (including, e.g., LFG groups) are lost -- this means at least four commands per reload (save, save_game, save_groups, reload) just to stop unreasonably penalizing some user
3. "As a user," it is strange and unfriendly to, once I have been informed of an unintentional abuse of the system, continue to be constrained by a penalty even after a period far longer than the penalty itself; e.g. if I have not willfully abused the system in 10 minutes (or worse, an hour), it seems slightly hostile to continue to be penalized with a grace period of 20 seconds. 
4. The spam guard is not concerned with usage patterns, only what user is attempting to run what count of commands (with no concept of relation) over some period of time. There are completely valid usage patterns that will be unfairly caught, e.g.:

!listSessions
(only 1 session is sent back, and it's the one a user will want; the command merely serves to confirm session ID)
!joinSession 1 (within 10 seconds)

The session would not be joined and the user would be penalized. Maybe the user wants to see the members of a group and then their credentials using bot commands; this would also trip the spam guard and cause further penalty. 

Possible remediations:

- It might be desirable to, at least for the sake of the channel, distinguish between public channel messages and private messages in the spam guard, so that you're not penalized for high command volume if no one but the bot and your own PM dialog is affected.
- It might also be desirable to establish certain "classes" of command for the purpose of flood filtering; a mechanism by which you define "you can only run 3 session-list commands within 10 seconds" or "you can only run 10 destiny.py commands within 10 seconds", such that each class may be disparate from one another and measured separately so as not to penalize the user for running otherwise dependent (or reasonably independent in some ways) commands. And maybe a "global" "class" to prevent overall flooding set reasonably high (higher than the others combined, at least, most likely) 
- A periodic task that simply resets penalties as is reasonable. (Preferably in conjunction with other suggestions) -- possibly commands to allow manual/administrative intervention for same